compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -1525,7 +1525,9 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
 
                 expr = borrowed;
 
-                if maybe_suggest(suggested_ty, count, suggestions.clone()) {
+                if self.tcx.sess.source_map().span_to_snippet(span).unwrap().starts_with("&")
```

Comment:
Could you move this check a bit higher and avoid doing `suggestions.push` if the snippet doesn't start with `&`?

Also, unwrap here can actually panic, the logic should be `if let Some(snippet) = ... && snippet.starts_with("&")`.